src/Sylius/Bundle/CoreBundle/Sitemap/Provider/ProductUrlProvider.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+ 
+namespace Sylius\Bundle\CoreBundle\Sitemap\Provider;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Sylius\Bundle\CoreBundle\Sitemap\Factory\SitemapUrlFactoryInterface;
+use Sylius\Bundle\CoreBundle\Sitemap\Model\SitemapUrlInterface;
+use Sylius\Component\Core\Model\Product;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * @author Arkadiusz Krakowiak <arkadiusz.krakowiak@lakion.com>
+ */
+class ProductUrlProvider implements UrlProviderInterface
+{
+    /**
+     * @var RepositoryInterface
+     */
+    private $repository;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var SitemapUrlFactoryInterface
+     */
+    private $sitemapUrlFactory;
+
+    /**
+     * @var array
+     */
+    private $urls;
+
+    /**
+     * @param RepositoryInterface $repository
+     * @param RouterInterface $router
+     * @param SitemapUrlFactoryInterface $sitemapUrlFactory
+     */
+    public function __construct(
+        RepositoryInterface $repository,
+        RouterInterface $router,
+        SitemapUrlFactoryInterface $sitemapUrlFactory
+    ) {
+        $this->repository = $repository;
+        $this->router = $router;
+        $this->sitemapUrlFactory = $sitemapUrlFactory;
+        $this->urls = array();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generate(array $criteria = array())
+    {
+        $products = $this->repository->findBy($criteria);
+
+        foreach ($products as $product) {
+            $productUrl = $this->sitemapUrlFactory->createEmpty();
+            $localization = $this->router->generate($product, array(), true);
+
+            $productUrl->setLastModification($product->getUpdatedAt());
+            $productUrl->setLocalization($localization);
+            $productUrl->setChangeFrequency(SitemapUrlInterface::CHANGEFREQ_ALWAYS);
+            $productUrl->setPriority(0.5);
+
+            array_push($this->urls, $productUrl);
```

Comment:
Since you're pushing only one item per time, would be better to just use
`$this->urls[] = $productUrl` IMO, in this case does the same thing.